### DISCUSSION TITLE
I am trying to understand the Magma architecture.

### DISCUSSION THREAD

**Discussion by bensternthal** [magma/magma#15205] [Q&A] [answered]
_This thread was copied over from [Slack](https://magmacore.slack.com/archives/C0191LLL03S/p1684829127479239)_

**Coderx0**
I am trying to understand the Magma architecture. In one of the documentation figures, we can see that the federation gateway connects to the HSS via the interface S6a. According to the 3GPP standard, the MME communicates with the HSS via the S6a. Does that mean that the FEG has an embeded MME inside ? If this is the case, where is it situated in the Magma code ?
Thank you.

**Phil Ritter**
I'll take a shot at it even though I am not really involved with the project any longer.  Its actually a pretty easy answer...

Magma does not have an identifiable MME, per se.  The Magma architecture implements the "core" functions of mobility as a single package in the Magma Access Gateway.  The "functions of" the MME, SGW, PGW and part of the HSS are all embedded in the AGW (with part of the HSS in the Orchestrator).  They are not really identifiable as separate functional units. The functional boundaries between them and the interfaces used to communicate between them were implemented in a way that Magma's original architects saw as being more aligned with how hyperscale systems are designed.

Similarly, for 5g, the AMF, SMF & UPF functions are implemented together as a single unit inside the AGW, albeit they are a bit more identifiable in the code as the 5g design is a bit more modern than 4g was.

The FEG exists to translate Magma's internal messaging format to more standard 3gpp interfaces when interaction with a more standard 3gpp network is required.  For subscriber functions, the FEG "presents" the S6a interface to the outside system and then communicates the required state information to the Orchestrater, which then relay's it to the AGW.

See here: https://magmacore.org/learn-the-technology/

**Coderx0**
Concerning the HSS, it is still unclear to me how it functions. What I understood from your previous explanation is that the HSS is divided into two parts: one in the orchestrator, and the other inside the AGW. If I append the network with an external HSS (for instance OpenAir HSS) that I connect to the FEG using the S6a interface, would that disable the HSS in magma (inside the orchestrator and AGW) ?

**Comment by Radula76** [comment, marked as answer]
Hi, the magma documentation goes over the architecture, start with:
https://magma.github.io/magma/docs/basics/introduction and
https://magma.github.io/magma/docs/lte/architecture_overview

your first question, the FEG doesn't have an MME inside it, its just forwarding the MME's s6a request to another HSS.
you can see the MME functions in magma in the above 2 links.

HSS is the repository/database for the subscribers.  Orc8r hosts the configuration part, where you can define or remove subscribers.  this data gets pushed down to each AGW on the network, so the AGW's have the HSS inside them.
If you configure an external HSS and use the FEG to forward the queries the AGW's will just use that instead.  it won't delete the HSS functionality in the Orc8r or AGW, it just won't use them.  You will need to manually(or whatever tools that HSS has) configure the external HSS with your subscriber profiles.

much more information here:
https://portal.3gpp.org/desktopmodules/Specifications/SpecificationDetails.aspx?specificationId=1712

or a breakdown of the HSS purpose, subscriber authentication here;
https://nickvsnetworking.com/hss-usim-authentication-in-lte-nr-4g-5g/
